### ISSUE TITLE
[design] Specializer design

### ISSUE THREAD

**Issue by mydmdm** [microsoft/SparTA#2]
*updated*

SparTA specializer generates customized codes for the sparse operators. 

***Highlight difference with existing implementation***
1. use [jinja2](https://jinja.palletsprojects.com/en/3.1.x/api/#) to handle templates
1. use [PyTorch JIT Cpp extension](https://pytorch.org/tutorials/advanced/cpp_extension.html#jit-compiling-extensions) to compile generated kernel codes, load kernel function, and test via numpy (torch) tensor
2. decouple `TeSA` transform (e.g., block coverage) and computing code emitter
3. definition of `TeSA` class, make it extensible for quantization formats and diverse storage layout (e.g., CSR, BCSR, etc.)
1. tune the performance in op level, other than in kernel function level
4. make the sparse operator class clean for example, let external tuner hold tuning history

***Known limits***
1. only support single transform and emitter in existing base class; we may need a nested search space and decorate to define a pipeline with multiple transform or emitter choices 
1. not support use sparse operator as submodule yet, but I believe current interface is compatible (that's why I use op-level tuning)
1. simplify the burden of tuner, we could support value choice (selecting from a list or set) only

## Sparse operators based on single TeSA transformer and emitter

SparTA offers some basic sparse operators e.g., sparse linear, sparse multi-head self-attention. These operators are based on the efficient sparse computing library (template). 

```python
from torch.utils.cpp_extension import load

__kernel_binding__ = None # must run op.specialize() first

class DynamicSparseLinearFunction(torch.autograd.Function):
    @staticmethod
    def forward(...): 
        return __kernel_binding__.forward_func(...)
    @staticmethod
    def backward(...):
        return __kernel_binding__.backward_func(...)

class DynamicSparseLinear(nn.Linear, SparseOpBase):

    def __init_specialization__(self, ...):
        self.tesa_transform = BlockCoverage
        self.emitter = TemplateEmitterBase(templates=['operators/dynamic_sparse_linear.cu', 'kernels/sparta_sparse_matmul.cu'])
        self.search_space = SearchSpace({ # we could support only value choice first
            'transform_params': ValueChoice((32,32), (32,64), (64,32), (64,64)), # value choice is a list / set of 
            'emitter_params': SearchSpace({
                'BLOCK_SIZE_M_VALUE': ValueChoice(8,16,32,64,128),
                #...
            })
        })

    def forward(self, ...):
        return DynamicSparseLinearFunction.apply(...)
    
    def specialize(self, tesa: dict, sample: TemplateParam):
        transform_params, emitter_params = ..(sample)..
        self.weight_tesa = self.tesa_transform(tesa['weight'], **transform_params)
        # meta information such as BCSR format (block size), number of non-zero block, indexes (even bidirectional) are embedded
        emitter_params.update(...) # update according to tesa
        files = self.emitter.code_gen(emitter_params)
        __kernel_binding__ = load(name, sources = files)
        return self

    def performance_optimization(self, tesa: dict, tuner_cls: type(Tuner), evaluator: Evaluator):
        tuner = tuner_cls(self.search_space)
        while not tuner.completed():
            samp = tuner.get_param()
            self.specialize(tesa, samp)
            metric = evaluator(self) # forward or backward step latency
            tuner.receive_metric(metric)
        return self.specialize(tesa, tuner.best_param()) 

class SparseWeightLinearEvaluator:

    def __init__(self, in_features, out_features):
        self.ref_op = nn.Linear(in_features, out_features)
        self.inputs = ...

    def evaluate(self, spop: SparseWeightLinear):
        pass

# usage example
weight_mask = ..Tensor..
weight_tesa = TeSAUtils.from_mask(weight_mask)
splinear = SparseWeightLinear(in_dim, out_dim)
splinear.tune_specialize({'weight': weight_tesa}, RandomSearchTuner, SparseWeightLinearEvaluator().evaluate)
```

## TeSA and transform

```python
@dataclass
class TeSA:
    mask: torch.Tensor # int tensor
    formats: List[str|dict]
    layout: dict = None # or name it as storage

class TeSAUtils
    @staticmethod
    def from_mask(mask: torch.Tensor, formats: list = None):
        return TeSA(mask=mask, layout={'layout': 'dense', 'meta': None},
                    formats=['float32', 'pruned'] if formats is None else formats)
```

The `mask` is integer tensor indicate each element's format, its value is between `0` and `#formats - 1`. 
The `formats` is the registered formats corresponding to the integer values in `mask`, for example, the common setting for pruning and quantization could be like 
```python
pruning_tesa = TeSA(..mask{0,1}.., ['float32', 'pruned'])
pruning_quant_tesa = TeSA(..mask{0,1}.., [{'format': 'int8', 'scale':.., 'zero-point': 0}, 'pruned'])
```
The `layout` is `{'layout': 'dense', 'meta': None}` by default, which means the tensor is stored in dense format. After transformation (or convert), it could be changed to a detailed sparse format like 
```python
{
    'layout': 'BCSR',
    'meta': {
        'block': (32,32),
        'index': ..numpy.array.., # optional
        'transpose_index': ..numpy.array.., # optional
    } 
}
```

## Emitter
The `emitter` class generate kernel codes for specific functions, e.g., sparse matrix multiplication, from a given template or scheduling primitives. 

Here is an example of template based emitter (via `jinja2`).
```python
from jinja2 import Template

class TemplateEmitterBase:

    def __init_template__(self, template_names: str, compile_opts: List[str]):
        self.templates = [t:Template(..path/to/template_name..) for t in template_names]
        self.compile_opts = compile_opts

    def code_gen(self, params: dict):
        src_code = [k:v.render(*params) for k,v in self.templates]
        files = .. write to disk ..
        return files
```

In fact we have two kinds of templates, the *computing kernels* and the *operators interface*. We could separate them by folders like:
```bash
templates
    |- kernels
        |- sparta_sparse_matmul.cu
        |- ...
    |- operators
        |- sparse_weight_linear.cu
        |- sparse_msa.cu
```

We could leverage the template engine to write complicated templates, for example, matmul with or without bias input as below
```jinja2
__global__ void BLOCK_SPARSE_MATMUL_BIAS(
    float* A, float* W_val, int* W_row, int* W_col,
    float* C, 
{% if USE_BIAS %}
    float *bias, 
{% endif %}
    int M, int K, int N){
        ...
    }
```

or we could let the template selecting different computing kernels according to its block size
```jinja2
void dynamic_forward_function(float* activation, int* row_ptr, int* col_idx,
                    float * val, float* bias, int M, int K, int N, int block_h, int block_w, float* output)
{
    dim3 gridDim(N/{{ BLOCK_SIZE_N }}, M/{{ BLOCK_SIZE_M }});
    dim3 blockDim({{ BLOCK_SIZE_N }} / {{ THREAD_SIZE_N }}, {{ BLOCK_SIZE_M }}/{{ THREAD_SIZE_M }});

    {% if (BLOCK_SIZE_N == 32) and (BLOCK_SIZE_M == 32) %}
    ...efficient 32x32 kernel...
    {% else %}
    ...general kernel...
    {% endif%}
}
```
